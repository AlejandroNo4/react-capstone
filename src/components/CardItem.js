import React, { useEffect } from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import { cleanCardData } from '../actions/index';
import { fetchingCard } from './fetching';

const GameItem = ({ match }) => {
  const dispatch = useDispatch();
  const cardState = useSelector((state) => state.singleCardReducer);
  const {
    params: { itemId },
  } = match;
  const url = itemId;

  useEffect(() => {
    fetchingCard({ dispatch, url });
    return () => { dispatch(cleanCardData()); };
  }, []);

  if (cardState.loading || Object.keys(cardState.data).length === 0) {
    return <h2>loading</h2>;
  }
  if (cardState.error) {
    return <h2>{cardState.error}</h2>;
  }

  const card = cardState.data;

  return (
    <div>
      <img alt={card.name} src={card.imageUrl} />
      <h2>{card.name}</h2>
      <p>Artist:</p>
      <p>{card.artist}</p>
      <p>Original Type:</p>
      <p>{card.originalType}</p>
      <p>Power:</p>
      <p>{card.power}</p>
      <p>Rarity:</p>
      <p>{card.rarity}</p>
      <p>Text:</p>
      <p>{card.text}</p>
      <p>
        <Link to="/">go back!</Link>
      </p>
    </div>
  );
};

GameItem.propTypes = {
  match: PropTypes.objectOf(PropTypes.any).isRequired,
};

export default GameItem;
