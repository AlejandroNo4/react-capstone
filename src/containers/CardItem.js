import React, { useEffect } from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import { cleanCardData } from '../actions/index';
import { fetchingCard } from '../api/fetching';
import Loading from './Loading';
import logo from '../assets/magic-logo.png';
import notFound from '../assets/not-found.png';

const CardItem = ({ match }) => {
  const dispatch = useDispatch();
  const cardState = useSelector((state) => state.cardReducer);
  const {
    params: { itemId },
  } = match;
  const url = itemId;

  useEffect(() => {
    fetchingCard({ dispatch, url });
    return () => { dispatch(cleanCardData()); };
  }, []);

  if (cardState.loading) {
    return <Loading />;
  }
  if (cardState.error) {
    return <h2>{cardState.error}</h2>;
  }

  const card = cardState.data;

  return (
    <div>
      <nav className="d-flex nav-bar">
        <p>
          <Link to="/"><img alt="logo" className="nav-logo" src={logo} /></Link>
        </p>
      </nav>
      <div className="d-flex flex-column align-center card-info cards-container">
        <h2 className="title">{card.name}</h2>
        <div className="img-container">
          <div className={card.imageUrl === undefined ? '' : 'line-card-top'} />
          <img alt={card.name} src={card.imageUrl === undefined ? notFound : card.imageUrl} />
          <div className={card.imageUrl === undefined ? '' : 'line-card-bottom'} />
        </div>
        <div className="info-container">
          <p className="property">Artist:</p>
          <p className="content">{card.artist}</p>
          <p className="property">Original Type:</p>
          <p className="content">{card.originalType}</p>
          <p className="property">Power:</p>
          <p className="content">{card.power}</p>
          <p className="property">Rarity:</p>
          <p className="content">{card.rarity}</p>
          <p className="property">Text:</p>
          <p className="content">{card.text}</p>
        </div>
        <Link to="/" className="back-button">BACK TO LIST</Link>
      </div>
    </div>
  );
};

CardItem.propTypes = {
  match: PropTypes.objectOf(PropTypes.any).isRequired,
};

export default CardItem;
