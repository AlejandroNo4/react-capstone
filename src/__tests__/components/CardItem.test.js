import React from "react";
import { Provider } from "react-redux";
import { render, screen, cleanup } from "@testing-library/react";
import axiosMock from "axios";
import CardItem from "../../containers/CardItem";
import "@testing-library/jest-dom";
import { BrowserRouter } from "react-router-dom";
import store from "../../store";

afterEach(cleanup);

const renderWithRedux = (component) => ({
  ...render(
    <Provider store={store}>
      <BrowserRouter>{component}</BrowserRouter>
    </Provider>
  ),
});

let dataMockfn;

describe("Card Item component tests", () => {
  beforeEach(() => {
    dataMockfn = axiosMock.get.mockResolvedValueOnce({
      data: {
        card: {
          id: "12345",
          artist: "Volkan Baǵa",
          imageUrl:
            "http://gatherer.wizards.com/Handlers/Image.ashx?multiverseid=129465&type=card",
          name: "Angel of Mercy",
          originalType: "Creature - Angel",
          power: "3",
          rarity: "Uncommon",
          text: "Angel of Mercy enters the battlefield, you gain 3 life.",
        },
      },
    });
    const mockMatch = {
      isExact: true,
      params: { itemId: "8ac972b5-9f6e-5cc8-91c3-b9a40a98232e" },
      path: "/item/:itemId",
      url: "/item/8ac972b5-9f6e-5cc8-91c3-b9a40a98232e",
    };
    renderWithRedux(<CardItem match={mockMatch} />);
  });

  it("Renders with Redux", async () => {
    const cardWrapper = await screen.findByTestId("card-wrapper");
    expect(cardWrapper).toBeVisible();
  });

  it("Matches the snapshot", async () => {
    const cardWrapper = await screen.findByTestId("card-wrapper");
    expect(cardWrapper).toMatchSnapshot();
  });

  it("Renders the correct route for BACK button", async () => {
    const buttonBack = await screen.findByTestId("back-btn");
    expect(buttonBack).toHaveProperty("href", "http://localhost/");
    expect(buttonBack).not.toHaveProperty("href", "/");
  });

  it("Renders the correct route for HOME button", async () => {
    const buttonBack = await screen.findByTestId("home-btn");
    expect(buttonBack).toHaveProperty("href", "http://localhost/");
    expect(buttonBack).not.toHaveProperty("href", "/");
  });

  it("All the props are the fetched from api TITLE", async () => {
    const title = await screen.findByTestId("title");
    expect(title).toHaveTextContent("Angel of Mercy");
    expect(title).not.toHaveTextContent("");
  });

  it("All the props are the fetched from api ARTIST", async () => {
    const infoContainer = await screen.findByTestId("info-container");
    expect(infoContainer.childNodes[1]).toHaveTextContent("Volkan Baǵa");
    expect(infoContainer.childNodes[1]).not.toHaveTextContent("");
  });

  it("All the props are the fetched from api ORIGINAL TYPE", async () => {
    const infoContainer = await screen.findByTestId("info-container");
    expect(infoContainer.childNodes[3]).toHaveTextContent("Creature - Angel");
    expect(infoContainer.childNodes[3]).not.toHaveTextContent("");
  });

  it("All the props are the fetched from api POWER", async () => {
    const infoContainer = await screen.findByTestId("info-container");
    expect(infoContainer.childNodes[5]).toHaveTextContent("3");
    expect(infoContainer.childNodes[5]).not.toHaveTextContent("");
  });

  it("All the props are the fetched from api RARITY", async () => {
    const infoContainer = await screen.findByTestId("info-container");
    expect(infoContainer.childNodes[7]).toHaveTextContent("Uncommon");
    expect(infoContainer.childNodes[7]).not.toHaveTextContent("");
  });

  it("All the props are the fetched from api TEXT", async () => {
    const infoContainer = await screen.findByTestId("info-container");
    expect(infoContainer.childNodes[9]).toHaveTextContent(
      "Angel of Mercy enters the battlefield, you gain 3 life."
    );
    expect(infoContainer.childNodes[9]).not.toHaveTextContent("");
  });
});
