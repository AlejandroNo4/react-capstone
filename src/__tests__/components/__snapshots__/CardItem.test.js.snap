// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Item component tests Matches the snapshot 1`] = `
<div
  data-testid="card-wrapper"
>
  <nav
    class="d-flex nav-bar"
  >
    <p>
      <a
        data-testid="home-btn"
        href="/"
      >
        <img
          alt="logo"
          class="nav-logo"
          src="magic-logo.png"
        />
      </a>
    </p>
  </nav>
  <div
    class="d-flex flex-column align-center card-info cards-container"
  >
    <h2
      class="title"
      data-testid="title"
    >
      Angel of Mercy
    </h2>
    <div
      class="img-container"
    >
      <div
        class="line-card-top"
      />
      <img
        alt="Angel of Mercy"
        src="http://gatherer.wizards.com/Handlers/Image.ashx?multiverseid=129465&type=card"
      />
      <div
        class="line-card-bottom"
      />
    </div>
    <div
      class="info-container"
      data-testid="info-container"
    >
      <p
        class="property"
      >
        Artist:
      </p>
      <p
        class="content"
      >
        Volkan Ba«µa
      </p>
      <p
        class="property"
      >
        Original Type:
      </p>
      <p
        class="content"
      >
        Creature - Angel
      </p>
      <p
        class="property"
      >
        Power:
      </p>
      <p
        class="content"
      >
        3
      </p>
      <p
        class="property"
      >
        Rarity:
      </p>
      <p
        class="content"
      >
        Uncommon
      </p>
      <p
        class="property"
      >
        Text:
      </p>
      <p
        class="content"
      >
        Angel of Mercy enters the battlefield, you gain 3 life.
      </p>
    </div>
    <a
      class="back-button"
      data-testid="back-btn"
      href="/"
    >
      Back to List!
    </a>
  </div>
</div>
`;
